ui/components/src/ThresholdsEditor/ThresholdsEditor.tsx

```diff
@@ -0,0 +1,236 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { RefObject, useEffect, useRef, useState } from 'react';
+import produce from 'immer';
+import { FormLabel, IconButton, TextField, ToggleButton, ToggleButtonGroup, Typography } from '@mui/material';
+import DeleteIcon from 'mdi-material-ui/DeleteOutline';
+import PlusIcon from 'mdi-material-ui/Plus';
+import { Stack } from '@mui/system';
+import { ThresholdOptions } from '@perses-dev/core';
+import { InfoTooltip } from '../InfoTooltip';
+import { useChartsTheme } from '../context/ChartsThemeProvider';
+import { OptionsEditorGroup } from '../OptionsEditorLayout';
+import { ThresholdColorPicker } from './ThresholdColorPicker';
+
+interface ThresholdsEditorProps {
+  thresholds?: ThresholdOptions;
+  onChange: (thresholds: ThresholdOptions) => void;
+}
+
+export function ThresholdsEditor({ thresholds, onChange }: ThresholdsEditorProps) {
+  const {
+    thresholds: { defaultColor, palette },
+  } = useChartsTheme();
+  const defaultThresholdColor = thresholds?.default_color ?? defaultColor;
+
+  const [steps, setSteps] = useState(thresholds?.steps);
+  useEffect(() => {
+    setSteps(thresholds?.steps);
+  }, [thresholds?.steps]);
+
+  // every time a new threshold is added, we want to focus the recently added input
+  const recentlyAddedInputRef = useRef<HTMLInputElement | null>(null);
+  const focusRef = useRef(false);
+  useEffect(() => {
+    if (!recentlyAddedInputRef.current || !focusRef.current) return;
+    recentlyAddedInputRef.current?.focus();
+    focusRef.current = false;
+  }, [steps?.length]);
+
+  const handleThresholdChange = (e: React.ChangeEvent<HTMLInputElement>, i: number) => {
+    setSteps(
+      produce(steps, (draft) => {
+        const step = draft?.[i];
+        if (step) {
+          step.value = Number(e.target.value);
+        }
+      })
+    );
+  };
+
+  const handleThresholdColorChange = (color: string, i: number) => {
+    if (thresholds !== undefined) {
+      onChange(
+        produce(thresholds, (draft) => {
+          if (draft.steps !== undefined) {
+            const step = draft.steps[i];
+            if (step) {
+              step.color = color;
+            }
+          }
+        })
+      );
+    }
+  };
```

Comment:
I was wondering what the difference between `handleThresholdChange` vs `handleThresholdColorChange` was... I thought maybe `handleThresholdChange` was changing the entire object. Could do `handleThresholdValueChange` if that seems like an improvement 🤷‍♀️